R/drive_list.R

```diff
@@ -0,0 +1,206 @@
+#' List files on Google Drive
+#'
+#' @param path character, path where the Google drive files are
+#'   that you would like to list. Defaults to the "My Drive" directory.
+#' @param pattern character, regular expression of titles of
+#'   documents to output in a tibble.
+#' @param ... name-value pairs to query the API
+#' @param verbose logical, indicating whether to print informative messages
+#'   (default `TRUE`)
+#'
+#'   This will default to the most recent 100 files on your Google Drive. If you
+#'   would like more than 100, include the `pageSize` parameter. For example, if
+#'   I wanted 200, I would run `drive_ls(pageSize = 200)`.
+
+#' Helpful links for forming queries:
+#'   * <https://developers.google.com/drive/v3/web/search-parameters>
+#'   * <https://developers.google.com/drive/v3/reference/files/list>
+
+#' @return tibble containing the name, type, and id of files on your google
+#'   drive (default 100 files)
+#' @examples
+#' \dontrun{
+#' ## list user's Google Sheets
+#' drive_list(q = "mimeType='application/vnd.google-apps.spreadsheet'")
+#' }
+#'
+#' @export
+drive_list <- function(path = NULL, pattern = NULL, ..., verbose = TRUE){
+
+  folder <- NULL
+  if (!is.null(path)) {
+    folder <- get_leafmost_id(path = path)
+  }
+  x <- list(...)
+
+  #add default fields if null
+  if (is.null(x$fields)) {
+    x$fields <- paste0("files/", .drive$default_fields, collapse = ",")
+  }
+
+  #add folder if not null
+  if (!is.null(folder)){
+    q <- paste0("'", folder, "'", " in parents")
+    if (!is.null(x$q)){
+      x$q <- paste(x$q, "and", q)
+    } else {
+      x$q <- q
+    }
+  }
+
+
+
+
+  request <- build_request(endpoint = .drive$base_url_files_v3,
+                           token = drive_token(),
+                           params = x)
+  response <- make_request(request)
+  proc_res <- process_request(response)
+  req_tbl <- tibble::tibble(
+    name = purrr::map_chr(proc_res$files, "name"),
+    type = sub(".*\\.", "", purrr::map_chr(proc_res$files, "mimeType")),
+    folder_id = purrr::map_chr(
+      purrr::map(
+        proc_res$files,
+        "parents",
+        .null = NA
+      ),
+      1),
+    id = purrr::map_chr(proc_res$files, "id"),
+    gfile = proc_res$files)
+
+  if (is.null(pattern)){
+    return(req_tbl)
+  } else{
+    if (!(inherits(pattern, "character") & length(pattern) == 1)){
+      stop("Please update `pattern` to be a character string.")
+    }
```

Comment:
This should be done at the vveerryy beginning, before nothing with an API calls, for example.